tutorials/haskell/snap-web-app/tutorial.md

```diff
@@ -0,0 +1,482 @@
+[record syntax](https://en.wikibooks.org/wiki/Haskell/More_on_datatypes#Named_Fields_.28Record_Syntax.29):
+
+```haskell
+data Todo = Todo
+  { todoID      :: Maybe Int
+  , description :: Text
+  , completed   :: Bool
+  } deriving (Eq, Show, Read)
+```
+[Back in the Site module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Site.hs#L143),
+there were used a function called `checkMigrationOrMigrate`
+receiving a database connection as parameter, while initalizing our Todo App.
+This function verifies if `todos` table exists in the database and if not, then creates it.
+
+```haskell
+checkMigrationOrMigrate :: Connection
+                        -> IO ()
+checkMigrationOrMigrate conn = do
+  schemaCreated <- tableExists conn "todos"
+  unless schemaCreated $ createTodoTable conn
+```
+
+In order to achieve its purpose, `checkMigrationOrMigrate` relies in a couple of
+aditional functions like `tableExists`:
+
+```haskell
+tableExists :: Connection
+            -> String
+            -> IO Bool
+tableExists conn tblName = do
+  let query = [i| SELECT name
+                  FROM sqlite_master
+                  WHERE type='table' AND name=? |]
+  name <- listToMaybe . fmap rowToValue <$> quickQuery' conn query [toSql tblName]
+  return (name == Just tblName)
+  where
+    rowToValue :: [SqlValue] -> String
+    rowToValue (snippetBody:_) = fromSql snippetBody
+    rowToValue _ = error "Unable to read single value"
+```
+
+And `createTodoTable`:
+
+```haskell
+createTodoTable :: Connection
+                -> IO ()
+createTodoTable conn = do
+  let query = [i| CREATE TABLE todos (
+                  id INTEGER PRIMARY KEY,
+                  description TEXT,
+                  completed BOOLEAN)|]
+
+  quickQuery' conn query []
+  commit conn
+```
+
+The function `tableExists` may look kind of tricky at first sight. It runs a very specific query
+to know if a table exists and then it maps this particular function composition:
+`listToMaybe . fmap rowToValue` over the query result, which should be the table name
+in debate, to extract it enclosed as a `Maybe` data type. Then it applies
+pattern matching to check if the table name is the expected one.
+
+I guess you noticed the use of
+[`quickQuery'`](http://hackage.haskell.org/package/HDBC-2.4.0.1/docs/Database-HDBC.html#v:quickQuery-39-)
+This function allows you to prepare, execute and fetch results in a more compact and idiomatic way.
+You'll see it around a lot.
+
+## Storing information in a database
+
+In order to deliver the expected features mentioned earlier the application have
+to perform the folllowing operations with Todo items againts a database:
+
+ - create
+ - update
+ - obtain
+ - delete
+ - list
+
+So, what happen if we map those operations over some code?😜. Well, the result
+are the following functions from the
+[Db module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Db.hs)
```

Comment:
@nebtrx 

"What happen**s**"

At the end, there's something missing. A period or  a colon, I'm not sure.